compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+                addFailureInstruction(errorMsg)
+                reporter.echo(errorMsg)
+                reporter.flushToFile()
+              }
+
+              acc + reporter.errorCount
+            }
+
+            completeCompilation(errorCount)
+
+            if (errorCount > 0) failTarget(target)
+          }
+        }
+
+      }
+    }
+  }
+
+  private final class RunTest(targets: List[Target], times: Int, threadLimit: Option[Int], suppressAllOutput: Boolean)
+  extends Test(targets, times, threadLimit, suppressAllOutput) {
+    private def verifyOutput(checkFile: JFile, dir: JFile, target: Target, warnings: Int) = try {
+        // Do classloading magic and running here:
+        import java.net.{ URL, URLClassLoader }
+        import java.io.ByteArrayOutputStream
+        val ucl = new URLClassLoader(Array(dir.toURI.toURL))
+        val cls = ucl.loadClass("Test")
+        val meth = cls.getMethod("main", classOf[Array[String]])
+
+        val printStream = new ByteArrayOutputStream
+        Console.withOut(printStream) {
+          meth.invoke(null, Array("jvm")) // partest passes at least "jvm" as an arg
+        }
+
+        val outputLines = printStream.toString("utf-8").lines.toArray
+        val checkLines = Source.fromFile(checkFile).getLines.toArray
+
+        def linesMatch =
+          outputLines
+          .zip(checkLines)
+          .forall { case (x, y) => x == y }
+
+        if (outputLines.length != checkLines.length || !linesMatch) {
+          // Print diff to files and summary:
+          val diff = outputLines.zip(checkLines).map { case (act, exp) =>
+            DiffUtil.mkColoredCodeDiff(exp, act, true)
+          }.mkString("\n")
+          val msg = s"\nOutput from run test '$checkFile' did not match expected, output:\n$diff\n"
+          echo(msg)
+          addFailureInstruction(msg)
+
+          // Print build instructions to file and summary:
+          val buildInstr = target.buildInstructions(0, warnings)
+          addFailureInstruction(buildInstr)
+
+          // Fail target:
+          failTarget(target)
+        }
+    }
+    catch {
+      case _: NoSuchMethodException =>
+        echo(s"\ntest in '$dir' did not contain a main method")
+        fail()
+
+      case _: ClassNotFoundException =>
+        echo(s"\ntest in '$dir' did was not contained within a `Test` object")
+        fail()
+
+      case _: InvocationTargetException =>
+        echo(s"\nTest in '$dir' might be using args(X) where X > 0")
+        fail()
+    }
+
+    protected def compilationRunnable(target: Target): Runnable = new Runnable {
+      def run(): Unit = compileTry {
+        val (errorCount, warningCount, hasCheckFile, doVerify) = target match {
+          case ConcurrentCompilationTarget(files, flags, outDir) => {
+            val sourceFiles = files.filter(isCompilable)
+            val checkFile = files.flatMap { file =>
+              if (file.isDirectory) Nil
```

Comment:
Because by convention, directories within directories compiled with `compileFilesInDir` are treated as `SeparateCompilationTarget`s. So it is taken care of by the other `case`